syft/frameworks/torch/tensors/interpreters/paillier.py

```diff
@@ -0,0 +1,296 @@
+from syft.generic.tensor import AbstractTensor
+from syft.generic.frameworks.hook import hook_args
+from syft.generic.frameworks.overload import overloaded
+from syft.workers.abstract import AbstractWorker
+import syft as sy
+import numpy as np
+import torch as th
+
+
+class PaillierTensor(AbstractTensor):
+    def __init__(self, owner=None, id=None, tags=None, description=None):
+        """Initializes a PaillierTensor, whose behaviour is to log all operations
+        applied on it.
+
+        Args:
+            owner: An optional BaseWorker object to specify the worker on which
+                the tensor is located.
+            id: An optional string or integer id of the PaillierTensor.
+        """
+        super().__init__(id=id, owner=owner, tags=tags, description=description)
+        print("creating paillier tensor 2")
+
+    def encrypt(self, public_key):
+        """This method will encrypt each value in the tensor using Paillier
+        homomorphic encryption.
+
+        Args:
+            *public_key a public key created using
+                syft.frameworks.torch.he.paillier.keygen()
+        """
+
+        output = PaillierTensor()
+        output.child = self.child
+        output.encrypt_(public_key)
+        return output
+
+    def encrypt_(self, public_key):
+        """This method will encrypt each value in the tensor using Paillier
+        homomorphic encryption.
+
+        Args:
+            *public_key a public key created using
+                syft.frameworks.torch.he.paillier.keygen()
+        """
+
+        inputs = self.child.flatten().tolist()
+        new_child = sy.pool().map(public_key.encrypt, inputs)
```

Comment:
I'm probably out of my depth here, since I don't yet know much about how this library works and how it's used, but this can be made considerably faster by precomputing the results of [obfuscate](https://github.com/data61/python-paillier/blob/955f8c0bfa9623be15b75462b121d28acf70f04b/phe/paillier.py#L596) (maybe in the background, in a separate thread / service) and then passing them to the `r_value` parameter of [encrypt](https://github.com/data61/python-paillier/blob/955f8c0bfa9623be15b75462b121d28acf70f04b/phe/paillier.py#L144). Even so, [this operation](https://github.com/data61/python-paillier/blob/955f8c0bfa9623be15b75462b121d28acf70f04b/phe/paillier.py#L136) should be precomputed as well, but it looks like the library was not designed that way.

Reply:
Agreed,
is this more a python-paillier issue or a Pysyft issue?

Reply:
I'd say both, given the limited design of `python-paillier`, but it depends on the intended usage of PySyft. If PySyft needs to be as fast as possible when calling the python-paillier `encrypt` function, then this is one way to speed it up considerably, at the expense of pre-computing randomisers separately in advance. Of course, this would have to be implemented in PySyft somehow.

Reply:
That's a great future work item! :)